### ISSUE TITLE
simple formatter not working in spots

### ISSUE THREAD

**Issue by jtrutwin** [mikeric/rivets#527]
Pretty new to rivetjs, I have a strange issue, that boils down to this:

```
  eventBind = rivets.bind(
    document.querySelector("#events-list"), {
      data: eventData
    }
  );

  rivets.formatters.length = function(value) {
    return value ? (value.length || 0) : 0;
  };
```

When I attempt to use the length formatter, I can use it within an rv-each, but not elsewhere within the bound element:

The filter works within the rv-each below, but not in the header th row:

```
<table id="events-list">
  <tr>
    <th>Sch <span rv-text="'foo' | length"></span></th>
            snip...
  </tr>
  <tr rv-each-event="data.events">
    <td>{event.schedule} <span rv-text="'foo' | length"></span></td>
           snip
```

In the th it shows "foo" - in the td within the rv-each it shows "3".

Is there a way to get this to work?  I need to use these formatters for other uses outside of rv-each.

Let me know if you need access to more information or the full code.

Thanks!

Josh

**Comment by Duder-onomy** [comment]
Hi Josh, 

I am a little confused. I made a working example from what I thought was your question. 
http://jsfiddle.net/0qt69bbg/
Can you take a look and let me know what is diff from your code?

If this resolves your issue please close. 
Greg

**Comment by jtrutwin** [comment]
Strange, looks like what I'm doing - I will try to post what I'm doing somewhere.

**Comment by jtrutwin** [comment]
Rather than try to re-create this, I'm just going to link to the site itself:

http://americantime.fivetechdev.com/events.htm

The source for that is here:

http://pastebin.com/8DQd7S10

If you search the source for "foo" you'll see the two places I use it and that the word "foo" is showin up in the URL above in the th tag, but not in the rv-each where the "3" is showing up in the Sch column.

The events.js file has the rivet binding code - but here that is for reference:

```
var eventData = { 
  events: [],
  schedules: [],
  scheduledata: [],
  schedule_ids: [],
  edit_event: { },
  
  rvFullDate: function() {
    console.log("rvFullDate", this);
    var year = this.get('year');
    var month = this.get('month');
    var day = this.get('day_of_month');
    
    return month + '/' + day + '/' + year;
  }
};

var eventBind;
var scheduleBind;
var eventEditBind;

function rvBindEvents() {  
  rivets.configure({
    adapter: {
      subscribe: function(obj, keypath, callback) {
        event = keypath ? 'change:' + keypath : 'change'
        obj.on(event, callback)
      },
      unsubscribe: function(obj, keypath, callback) {
        event = keypath ? 'change:' + keypath : 'change'
        obj.off('change:' + keypath, callback)
      },
      read: function(obj, keypath) {
        return keypath ? obj.get(keypath) : obj
      },
      publish: function(obj, keypath, value) {
        obj.set(keypath, value)
      }
    }
  });

  eventBind = rivets.bind(
    document.querySelector("#events-list"), {
      data: eventData
    }
  );
  scheduleBind = rivets.bind(
    document.querySelector("#schedule_select"), {
      data: eventData
    }
  );
  eventEditBind = rivets.bind(
    document.querySelector("#modalEditForm"), {
      data: eventData
    }
  );

  rivets.formatters.eventHour2 = {
    read: function (val) {
      console.log("eventHour", val);
      return val.split(':')[0];
    },
    publish: function(val) {
      consol.log("eventHour publish");
    }
  };

  rivets.formatters.eventHour = function (val) {
    console.log("eventHour", val);
    return val.split(':')[0];
  };
  
  rivets.formatters.eventMinute = function (value) {
    return value.split(':')[1].replace(/am|pm/, "");
  };
  
  rivets.formatters.eventAMPM = function (value) {
    return value.split(':')[1].replace(/\d+/, "");
  };

  rivets.formatters.length = function(value) {
    return value ? (value.length || 0) : 0;
  };

  console.log(rivets);
  
  getEventTableData();
  getSchedulenamesData();
}
```

The getEventTableData and getSchedulenamesData calls at the end of this load the eventData object from XHR calls.

Sorry this is kind of a mess, curious if you see any reason why these formatters are behaving like this?

Josh

**Comment by Duder-onomy** [comment]
Hi Josh, 

I looked through your code and everything seems reasonable (cept that angular reference, much shame)(...kidding)( or am i? )

I see 2 things that could be the issue. (besides the angular ref)

1. The way you define your adapter is incorrect. 
    In your code, I can only find references to Rivets binding paths using dot notation. Rivets ships with a 'dot' notation adapter, so you do not need to define a new one. Creating your own adapter would be useful to you if you were trying to bind something besides POJO's, like Backbone Models or really anything that you find useful in your domain. The adapter is used to teach Rivets how to observe your bound objects for changes. I have in the past, for example, I have written adapters to work over web sockets.  In your case, you do not need to define an adapter, so just delete that code. That being said, the way you defined your adapter is incorrect. Please look [here](http://rivetsjs.com/docs/guide/#adapters-creating)  The important thing to note as being diff from your example is that you must define the 'key' that the adapter will use. This 'key' tells rivets that, "for this keypath that is being bound, use this adapter"  Please let me know if you need additional information on that one. 

2. You define your custom binders and formatters after you bind. 
    Calls to `Rivets.bind()` will run through the bound object, looking at all the bindings, formatters, etc in dom node that you pass to it, building up a big old object which rivets will use to do its thing. So, if the custom formatters are not present at the initial bind, then rivets will not know about them, and thus, ignore them when building up its state. So, you should ensure that you add all of your formatters and binders `before` you call `rivets.bind`.  The reason that the ` | length` formatter you created worked later in your code is because the second usage of that formatter is inside a `rv-each`, at the point the children are bound, the `rivets.formatters.length` has been read in. 

Please let us know if this helps you. 
Greg

**Comment by jtrutwin** [comment]
Thanks, that helps.  I copied that adapter code verbatim from some other site, they have it wrong too I guess.  

And as for #2, that seems to have fixed the problem.  I swear I had tried that before but because my formatters weren't checking for value being undefined there would be an error on the eventHour formatter, when I moved it under the bind the error was gone.  I modified the formatters to check for value.   It'd be nice to have some documentation that formatters should be above rivets.bind() on http://rivetsjs.com/docs/guide/#formatters   :)

Thanks for the help, closing the ticket.  Is this the best place for a newb like me to get assistance?  

Josh
